### PR TITLE
Split bitbucket support into bitbucketCloud and bitbucketServer

### PR DESCRIPTION
**Pull request by pjungermann**
## Hey, I just made a Pull Request!

This PR related to #9923 contains multiple commits which will separate Bitbucket Cloud and Bitbucket Server
support.

Each commit is self-contained and could be extracted into a separate PR if this is desired.

The only aspect not included is the split of the processor. This is intentional as also written at the closing comment at PR #10611.

I have prepared an entity provider for Bitbucket Cloud related to #10183 which I would base on the first commit of this PR/the introduced `integrations.bitbucketCloud`. Users transitioning to the entity provider would already migrate to both in one go. It will also allow some further improvements to the communication with the external API.

This PR focuses on the split. Certain improvements to the different parts might be possible, but rather as separate changes.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .changeset/moody-suns-smell.md
Comment (Rugvip) [inline .changeset/moody-suns-smell.md:36]: Wanna bounce some ideas around this to make sure we're not complicating things unnecessarily. This is ofc a pretty big change so wanna tread carefully.

Am I right in assuming that Bitbucket Cloud is only available on `bitbucket.org`? Or can you have an on-prem deployment of that too?

The main concern is whether the config split is necessary or if it should be done in a slightly different way. The three other approaches I can think of are:

```yaml
integrations:
  bitbucket:
    - host: bitbucket.org # Implicit type=cloud because bitbucket.org
      username: bitbucket_user
      appPassword: app-password
    - host: bitbucket-server.company.com
      token: my-token
```

```yaml
integrations:
  bitbucket:
    - host: bitbucket.org
      type: cloud # explicit type
      username: bitbucket_user
      appPassword: app-password
    - host: bitbucket-server.company.com
      type: server
      token: my-token
```

```yaml
integrations:
  bitbucket:
    - username: bitbucket_user
      appPassword: app-password
  bitbucketServer: # leave one of them as is and let the other use the separate integration type
    - host: bitbucket-server.company.com
      token: my-token
```

With all of there we'd have a pretty much complete separation of implementation with early type checks everywhere and adding a `type` to the Bitbucket config object.

Gotta say that looking at the problem it really seems sane to have them be completely separated like you suggest though, weird and unfortunate but perhaps necessary.

As a bit of a separate note, are the capabilities of the APIs very different as well? I'm thinking for things like the scaffolder action, whether it makes sense to split or abstract there. I'd say that would largely depend on whether the options are completely different between the two.

**Review by Rugvip — APPROVED** [review]
Awesome stuff! :grin: 🎉 

Wanna start a bit high-level just to get that out of the way first. Sorry if I missed parts of the discussion up until this point, lots going on 😅

File: packages/backend-common/src/reading/BitbucketUrlReader.ts
Comment (Rugvip) [inline packages/backend-common/src/reading/BitbucketUrlReader.ts:50]: Another discussion for a different thread that I pretty arbitrarily chose to but here: I'm noticing this is pretty much only additions. Would it be possible to express more of the existing and deprecated Bitbucket implementations in terms of the two new ones? For example having the existing `BitbucketUrlReader` leverage the two new readers instead of having its own separate implementation?

If that just complicates things unnecessarily I think we should be pretty aggressive about the deprecations so that we can get the old code sooner rather than later. I'm thinking things like logging a warning when `bitbucket` is used.

File: .changeset/moody-suns-smell.md
Comment (pjungermann) [reply .changeset/moody-suns-smell.md:36]: Yes, bitbucket.org is the only host for Bitbucket Cloud.  The setup is a bit complicated with Bitbucket Cloud, Bitbucket Server (on-prem) and Bitbucket (Server) Data Center (managed Bitbucket Server). Bitbucket Server was formally known as Stash and was just rebranded after Atlassian bought Bitbucket (Cloud).

Comment (pjungermann) [reply .changeset/moody-suns-smell.md:36]: In general, I like `bitbucket` + `bitbucketServer`, too, but makes it a bit hard for the migration and due to it being used at so many places.

Using `bitbucketCloud` + `bitbuckerServer` seemed easier for splitting it and more explicit. Unfortunately, it is a bit set up for confusion..

Theoretically, we could stick to `integrations.bitbucket` only, but that might be misaligned with other parts of the configuration.

There is an auth provider for Bitbucket Cloud using `bitbucket` though. But also here, many people asked whether this would support Bitbucket Server as well -- even though there is a warning at the documentation about it.

Comment (pjungermann) [reply .changeset/moody-suns-smell.md:36]: Re API: yes, they are different and the reason why there were so many "isBitbucketCloud" checks at the code.

I've split the client at PR #10515

File: packages/backend-common/src/reading/BitbucketUrlReader.ts
Comment (pjungermann) [reply packages/backend-common/src/reading/BitbucketUrlReader.ts:50]: logging a warning sounds like a great idea! It might be possible to reuse the new ones at the old one. I can look into it.

Comment (pjungermann) [reply packages/backend-common/src/reading/BitbucketUrlReader.ts:50]: ah.. reusing will be tricky due to the selection of a ready by host. But I will add the warning.

Comment (pjungermann) [reply packages/backend-common/src/reading/BitbucketUrlReader.ts:51]: I added warnings at the UrlReader (and therefore, implicitly the integration) as well as the scaffolder action.

File: .changeset/moody-suns-smell.md
Comment (pjungermann) [reply .changeset/moody-suns-smell.md:36]: overall, I think it would be great to extract all of this into separate modules for bitbucket cloud and server incl. UrlReader, etc. So far, this is not yet possible though.

**Comment by pjungermann** [comment]
I just thought again about externalizing UrlReader implementations similar to what was done with the catalog-backend-module-{xyz} plugins.

This should be possible by not using UrlReaders.default but UrlReaders.create and pass them explicitly.

Currently, they are part of backend-common. Currently, SCM support is spread across different packages, etc. Probably, this still make sense (and bundling them somehow would be possible still, too, I guess). Maybe, something like backend-common-module-{aws|bitbucket-cloud|github|...}.

Currently, we still have a dependency to the aws-sdk, and similar due to the readers as they also need to communicate with external APIs, etc. This would allow to change this.

I bring this up here as it may make sense to kick this off with the new UrlReaders. Or maybe not and rather afterwards.

@Rugvip What do you think about this?

**Comment by Rugvip** [comment]
@pjungermann Yep we likely want to do something like that but I think we can hold off on it for now. It's very likely that we'll introduce a `backend-defaults` packages that ends up being the default setup similar to `app-defaults`, so you got the aggregation there. We just don't know exactly what kind of split we want yet I think, and what the APIs for bringing things back together will be.

**Review by Rugvip — APPROVED** [review]
Realized I approved it by mistake last time 😅, this one for real tho :p

@backstage/maintainers I'm thinking we're good to go here and getting it out early in 1.1 gives it some extra time for testing. The split is unfortunate but given the state of Bitbucket I'm feeling anything else is just a lot more confusing

**Comment by pjungermann** [comment]
an alternative to the split could have been to do a new integration "from scratch" already considering ideas like externalized UrlReader, etc.

**Comment by pjungermann** [comment]
Just to mentioned: I prepared a BitbucketCloudEntityProvider based on the first commit of this PR (integrations) and would file a PR for it after this one was merged.

**Comment by jhaals** [comment]
Thanks Patrick 👍

**Pull request merged**
